### PR TITLE
Support exits for NFTs minted on plasma

### PR DESCRIPTION
**Pull request by atvanguard**

### CODE REVIEW

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 1]
```diff
@@ -208,71 +211,49 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
...
+      // limit the gas amount that predicate.onFinalizeExit() can use, to be able to make gas estimations for bulk process exits
+      address exitor = ExitNFT(exitNFTContract).ownerOf(exitId);
+      uint256 amountOrNft = currentExit.receiptAmountOrNFTId;
+      address predicate = currentExit.predicate;
+      uint256 _gas = gasLimit - 52000; // sub fixed processExit cost , ::=> can't read global vars in asm
+      assembly {
+        let ptr := mload(64)
+        // keccak256('onFinalizeExit(address,address,uint256)') & 0xFFFFFFFF00000000000000000000000000000000000000000000000000000000
+        mstore(ptr, 0xfdd3d6bd00000000000000000000000000000000000000000000000000000000)
+        mstore(add(ptr, 4), exitor)
```
Comment (0xAshish) [inline contracts/root/withdrawManager/WithdrawManager.sol]: this must be payable!?

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 2]
```diff
@@ -208,71 +211,49 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
...
-      // Check if this exit has finished its challenge period.
-      if (exitableAt > block.timestamp) {
-        return;
-      }
-
-      // get withdraw block
+      // Stop processing exits if the exit that is next is queue is still in its challenge period
+      if (exitableAt > block.timestamp) return;
+      // If the exitNft was deleted as a result of a challenge, skip processing this exit
+      if (!exitNft.exists(exitId)) continue;
```
Comment (0xAshish) [inline contracts/root/withdrawManager/WithdrawManager.sol]: there exists `continue` 😄.

File: contracts/root/predicates/MarketplacePredicate.sol [hunk 3]
```diff
@@ -99,6 +102,12 @@ contract MarketplacePredicate is PredicateUtils {
     withdrawManager.addInput(priority /* exitId */, reference2.age /* age of input */, exitTxData.counterParty /* signer */);
   }
 
+  function onFinalizeExit(address exitor, address token, uint256 tokenId)
```
Comment (0xAshish) [inline contracts/root/predicates/MarketplacePredicate.sol]: only withdraw manager.

File: contracts/common/tokens/ERC721PlasmaMintable.sol [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+contract ERC721PlasmaMintable is ERC721Mintable {
+  /**
+   * @dev Returns whether the specified token exists
+   * @param tokenId uint256 ID of the token to query the existence of
+   * @return bool whether the token exists
+   */
+  function exists(uint256 tokenId) public view returns (bool) {
+    return _exists(tokenId);
+  }
+}
```
Comment (0xAshish) [inline contracts/common/tokens/ERC721PlasmaMintable.sol]: new line 🌚

File: contracts/root/predicates/ERC721Predicate.sol [hunk 5]
```diff
@@ -117,6 +124,29 @@ contract ERC721Predicate is IErcPredicate {
...
+    external
+  {
+    address rootToken = startExitWithBurntTokens(data);
+    RLPReader.RLPItem[] memory txList = mintTx.toRlpItem().toList();
+    (address minter,) = getAddressFromTx(txList, withdrawManager.networkId());
+    require(
+      ERC721PlasmaMintable(rootToken).isMinter(minter),
+      "Not authorized to mint"
+    );
+  }
```
Comment (0xAshish) [inline contracts/root/predicates/ERC721Predicate.sol]: mint here only to deposit manager.

File: contracts/child/ChildERC721Mintable.sol [hunk 6]
```diff
@@ -0,0 +1,10 @@
+pragma solidity ^0.5.2;
+
+import { ChildERC721 } from "./ChildERC721.sol";
+import { ERC721Mintable } from "openzeppelin-solidity/contracts/token/ERC721/ERC721Mintable.sol";
+
+contract ChildERC721Mintable is ChildERC721, ERC721Mintable {
+  constructor (address rootToken)
+    ChildERC721(msg.sender, rootToken, "Mintable 721", "M721")
```
Comment (jdkanani) [inline contracts/child/ChildERC721Mintable.sol]: Who will be the minter?

Comment (atvanguard) [reply contracts/child/ChildERC721Mintable.sol]: The authorized minters fn the root chain 721 contract! The minters on root will need to be added to `ChildERC721Mintable.addMinter(address account)` (mapped contract on child)

Comment (jdkanani) [reply contracts/child/ChildERC721Mintable.sol]: Got it.

File: contracts/root/predicates/ERC721Predicate.sol [hunk 7]
```diff
@@ -82,12 +87,37 @@ contract ERC721Predicate is IErcPredicate {
     withdrawManager.addExitToQueue(msg.sender, childToken, rootToken, tokenId, txHash, burnt, age);
   }
 
+  /**
+   * @notice Start an exit for a token that was minted and burnt on the side chain
+   * @param data RLP encoded data of the burn tx
+   * @param mintTx Signed mint transaction
+   */
+  function startExitWithBurntTokens(bytes calldata data, bytes calldata mintTx)
```
Comment (jdkanani) [inline contracts/root/predicates/ERC721Predicate.sol]: Can we change the name? Some libs give problems with overridden methods.

**Pull request merged**